drafts/2019-10-01-rewriting-irewritable.md

```diff
@@ -0,0 +1,448 @@
+
+```csharp
+class ThreeChildren : IRewritable<ThreeChildren>
+{
+    // ...
+    public ThreeChildren RewriteChildren(Func<ThreeChildren, ThreeChildren> f)
+        => new ThreeChildren(f(_child1), f(_child2), f(_child3));  // no ImmutableList
+}
+```
+
+The extra method makes `IRewritable` harder to understand and harder to implement. Implementing `RewriteChildren` (without using `DefaultRewriteChildren`) can be tricky; if the transformer function doesn't change anything you should avoid recreating the object.
+
+
+## The new way
+
+I'm pleased to say that `IRewritable`'s new design addresses all three of these warts. Remember the "buffer" idea from earlier?
+
+```csharp
+interface IRewritable<T> where T : IRewritable<T>
+{
+    int CountChildren();
+    void GetChildren(Span<T> buffer);
+    T SetChildren(ReadOnlySpan<T> newChildren);
+}
+```
+
+This design is almost identical to the one I outlined earlier --- it uses [`Span`](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-3.0)s instead of arrays. (For the uninitiated, a `Span` is basically a "slice" of an array. It can also be backed by _unmanaged_ memory, though, making it more flexible than `ArraySegment`.)
+
+* `CountChildren` tells Sawmill how much space is needed to copy the children.
+* `GetChildren` copies the current object's immediate children into `buffer`.
+* `SetChildren` creates a copy of the current object with its immediate children replaced.
+
+Here's the important difference. Unlike an array, a `Span` _can't be stored on the heap_. The compiler checks that `Span`s are always confined to the stack. If you pass a `Span` into a method as an argument, you can be confident that the `Span` won't leave the scope of that method's stack frame (just like `ref`).
+
+So `Span` thoroughly solves the safety issue with the array-oriented API. I don't have to worry about mutating a `Span` which got stored inside a client object, because the `Span` can't be stored! This allows Sawmill to safely reuse `Span`s, rather than allocating a new array for each `GetChildren`/`SetChildren` call.
+
+So Sawmill's methods like `RewriteChildren` can be implemented without allocating memory. In this example I'm using [`ArrayPool`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=netcore-3.0) to avoid creating a new array for every `RewriteChildren` call.
+
+```csharp
+public static T RewriteChildren<T>(this T value, Func<T, T> transformer) where T : IRewritable<T>
+{
+    var count = value.CountChildren();
+    var array = ArrayPool<T>.Shared.Rent(count);
```

Comment:
I'd use [`MemoryPool<T>` ](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.memorypool-1?view=netcore-3.0 )instead, as you can `using` the return of [`Rent`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.memorypool-1.rent?view=netcore-3.0#System_Buffers_MemoryPool_1_Rent_System_Int32_), letting you remove the call to `Return()` below.

Reply:
The disposables aren't reused, though: https://github.com/dotnet/corefx/blob/4a223a0bd4d539dce1b5eea3f7657ada5217baeb/src/System.Memory/src/System/Buffers/ArrayMemoryPool.cs#L22.